extra_data/reader.py

```diff
@@ -1315,6 +1315,138 @@ def keys_detail(s, keys, prefix=''):
 
         print()
 
+    def plot_missing_data(self, min_saved_pct=95, all_xtdf_subsections=False):
+        """Plot sources that have missing data for some trains.
+
+        Parameters
+        ----------
+
+        min_saved_pct: int or float, optional
+            Only show sources with less than this percentage of trains saved.
+        all_xtdf_subsections: bool, optional
+            Control which subsections of XTDF sources are checked. By default
+            only the ``image`` subsection will be checked.
+        """
+        n_trains = len(self.train_ids)
+
+        # Helper function that returns an alias for a source if one is
+        # available, and the source name otherwise.
+        def best_src_name(src):
+            for alias, alias_ident in self._aliases.items():
+                if isinstance(alias_ident, str) and alias_ident == src:
+                    return alias
+
+            return src
+
+        # Find sources with missing data
+        flaky_sources = { }
+        save_pcts = { }
+        run_tids = set(self.train_ids)
+        start = time.time()
+        for src in self.all_sources:
+            kds = { }
+            if src.endswith(":xtdf"):
```

Comment:
While looking through the implementation in the context of plotting data counts, I noticed I had missed this earlier.

Any instrument source can have multiple index groups (the first part of the key), and there are already two non-XTDF instances in calibration with this. For that purpose `SourceData.data_counts()` (and `SourceData.drop_missing_trains()`) has an optional `index_group` to distinguish between those.

Reply:
With xtdf detectors, trains with 0 data should have 0 across all index groups. Does this hold for the other cases? I.e. can we summarise the data presence for a source here, or do we need to split it out into N groups?

Reply:
It depends on how much behaviour of the current DAQ we want to encode. Purely from an EXDF point of view, any index group is separate, and any instrument source may have multiple index groups with differing counts.

In practical terms probably, but I honestly don't know.

* A Karabo pipeline can have multiple index groups, but as the DAQ only accepts a single and structurally identical hash for each train, they cannot differ (all index groups are 0 or 1). 
* With XTDF, the [protocol itself](https://docs.xfel.eu/share/page/site/IT/document-details?nodeRef=workspace://SpacesStore/69ddf412-b1cf-4a65-b715-39bbbef2c20f) does not seem to mandate a non-zero `pulseCount` field in the header and what happens with the `image` section in this case. We should verify with ITDM that the DAQ does not add any train entry (neither `header`, `detector`, `image` nor `trailer`) when there's no frame data for a particular train, or whether it depends (detector intentionally sending XTDF data with 0 frames, no XTDF data all, online data reduction with empty pattern, etc).

Naturally the virtual sources we create in corrections are an entirely different matter. At the moment the REMI case always has the same number of entries for its index groups, but that is definitely something data reduction may change.

Would this have a significant performance impact in the current version?

Reply:
I'm not so concerned about the performance impact, I'm thinking more of how the information is presented, and the complexity of the code. Multiplying the number of bars when the different groups within a source have the same trains missing can make it harder to get a useful overview - see the example James showed in [this comment](https://github.com/European-XFEL/EXtra-data/pull/402#issuecomment-1563539674).

We could load all the groups for a source, check if they match, and if so only draw one bar. But that starts to sound like a fair bit of complexity for a corner case that it sounds like we don't currently need.

Luckily the details of the plot aren't part of the API contract, so perhaps we can get away with doing something simple now and adding complexity later if necessary.